segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java

```diff
@@ -0,0 +1,943 @@
+import io.pravega.segmentstore.storage.metadata.StorageMetadataException;
+import io.pravega.shared.NameUtils;
+import lombok.Builder;
+import lombok.Data;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.Setter;
+import lombok.extern.slf4j.Slf4j;
+import lombok.val;
+import lombok.var;
+
+import java.io.ByteArrayInputStream;
+import java.io.EOFException;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static io.pravega.segmentstore.storage.metadata.StorageMetadata.fromNullableString;
+import static io.pravega.segmentstore.storage.metadata.StorageMetadata.toNullableString;
+
+/**
+ * This class implements system journaling functionality for critical storage system segments which is useful for bootstrap after failover.
+ * It records any layout changes to storage system segments.
+ * Storage system segments are the segments that the storage subsystem uses to store all metadata.
+ * This creates a circular dependency while reading or writing the data about these segments from the metadata segments.
+ * System journal is a mechanism to break this circular dependency by having independent log of all layout changes to system segments.
+ * Currently only two actions are considered viz. Addition of new chunks and truncation of segments.
+ * This log is replayed when the ChunkManager is booted.
+ * To avoid data corruption. Each instance writes to its own distinct log file.
+ * During bootstrap all the system journal files are read and processed to re-create the state of the storage system segments.
+ */
+@Slf4j
+public class SystemJournal {
+    /**
+     * Serializer for {@link SystemJournalRecordBatch}.
+     */
+    private static final SystemJournalRecordBatch.SystemJournalRecordBatchSerializer BATCH_SERIALIZER = new SystemJournalRecordBatch.SystemJournalRecordBatchSerializer();
+
+    /**
+     * Serializer for {@link SystemSnapshotRecord}.
+     */
+    private static final SystemSnapshotRecord.Serializer SYSTEM_SNAPSHOT_SERIALIZER = new SystemSnapshotRecord.Serializer();
+
+    private final Object lock = new Object();
+
+    @Getter
+    private final ChunkStorage chunkStorage;
+
+    @Getter
+    private final ChunkMetadataStore metadataStore;
+
+    /**
+     * Epoch of the current instance.
+     */
+    @Getter
+    private final long epoch;
+
+    /**
+     * Container id of the owner container.
+     */
+    @Getter
+    private final int containerId;
+
+    /**
+     * Index of current journal file.
+     */
+    @Getter
+    private int currentFileIndex;
+
+    /**
+     * String prefix for all system segments.
+     */
+    @Getter
+    @Setter
+    private String systemSegmentsPrefix;
+
+    /**
+     * System segments to track.
+     */
+    @Getter
+    @Setter
+    private String[] systemSegments;
```

Comment:
It's best if you keep this collection internal and add APIs to mutate it. Otherwise external classes will need to know the internals of this class which can lead to trouble.

Reply:
it was for tests. Fixed tests

Reply:
It still is exposed.

Reply:
No it is fixed https://github.com/pravega/pravega/pull/4686/commits/90c2dd483f8507bb02048504a166fe71ae6a82ae